### PR TITLE
pythonPackages.pystray: Add appindicator support

### PR DESCRIPTION
**Pull request by Myaats**
###### Description of changes

The pystray python package has three backends for Linux: appindicator, gtk and xlib. So far only xlib has been supported, but is broken with the appindicator extension on GNOME. This introduces the dependencies required to make it work.

As this is based on gobject-introspection applications depending on pystray needs to be wrapped using wrapGApps, in this case it was only `jellyfin-mpv-shim`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by jojosch — CHANGES_REQUESTED** [review]
Thank you for fixing the appindicator. `jellyfin-mpv-shim` now shows the tray icon and all its menu options work.

-----

I'm getting an error while running `nixpkgs-review` since `libayatana-appindicator-gtk3` is an [alias](https://github.com/NixOS/nixpkgs/blob/6a6b2482438c47c5af1bee65d1d86d854b480762/pkgs/top-level/aliases.nix#L725-L726) of `libayatana-appindicator` and `nixpkgs-review` runs with `allowAliases = false`:

```
error: Function called without required argument "libayatana-appindicator-gtk3" at .../.cache/nixpkgs-review/pr-199761-1/nixpkgs/pkgs/development/python-modules/pystray/default.nix:12
```

```diff
diff --git a/pkgs/development/python-modules/pystray/default.nix b/pkgs/development/python-modules/pystray/default.nix
index 39416d4c0fa..975f940da68 100644
--- a/pkgs/development/python-modules/pystray/default.nix
+++ b/pkgs/development/python-modules/pystray/default.nix
@@ -9,7 +9,7 @@
 , gobject-introspection
 , pygobject3
 , gtk3
-, libayatana-appindicator-gtk3 }:
+, libayatana-appindicator }:
 
 buildPythonPackage rec {
   pname = "pystray";
@@ -23,7 +23,7 @@ buildPythonPackage rec {
   };
 
   nativeBuildInputs = [ gobject-introspection sphinx ];
-  propagatedBuildInputs = [ pillow xlib six pygobject3 gtk3 libayatana-appindicator-gtk3 ];
+  propagatedBuildInputs = [ pillow xlib six pygobject3 gtk3 libayatana-appindicator ];
   checkInputs = [ xvfb-run ];
 
   checkPhase = ''
```

fixes that error. After that `nixpkgs-review` passes:

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>jellyfin-mpv-shim</li>
    <li>plex-mpv-shim</li>
    <li>python310Packages.pystray</li>
    <li>python39Packages.pystray</li>
  </ul>
</details>

**Comment by Myaats** [comment]
> I'm getting an error while running `nixpkgs-review` since `libayatana-appindicator-gtk3` is an [alias](https://github.com/NixOS/nixpkgs/blob/6a6b2482438c47c5af1bee65d1d86d854b480762/pkgs/top-level/aliases.nix#L725-L726) of `libayatana-appindicator` and `nixpkgs-review` runs with `allowAliases = false`:

Fixed and force pushed

**Comment by Myaats** [comment]
Just noticed `plex-mpv-shim` will need to be wrapped too, no idea why I did not find it during previous searches :upside_down_face:

**Comment by Myaats** [comment]
I was unable to test `plex-mpv-shim` as it seems to have an unrelated error from something related to shader packs, the same error also happens with a build form HEAD on my machine.

```
2022-11-06 00:33:25,073 [    INFO] conf: Loaded settings from json: /home/mats/.config/plex-mpv-shim/conf.json
2022-11-06 00:33:25,299 [    INFO] player: Using libmpv1 playback backend.
2022-11-06 00:33:25,305 [   ERROR] menu: Could not load profile manager.
Traceback (most recent call last):
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/menu.py", line 63, in __init__
    profile_manager = VideoProfileManager(self, playerManager)
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/video_profile.py", line 24, in __init__
    self.load_shader_pack()
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/video_profile.py", line 36, in load_shader_pack
    raise FileNotFoundError("Could not find default shader pack.")
FileNotFoundError: Could not find default shader pack.
Traceback (most recent call last):
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/bin/.plex-mpv-shim-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/mpv_shim.py", line 47, in main
    from .player import playerManager
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/player.py", line 568, in <module>
    playerManager = PlayerManager()
  File "/nix/store/8j8f30zr8i8771jlrw5csp7s18rwdfiy-plex-mpv-shim-1.10.3/lib/python3.10/site-packages/plex_mpv_shim/player.py", line 235, in __init__
    def handle_end_idle(event):
  File "/nix/store/zr87l99vadmr157sa0vbjvwb5mgkc65c-python3.10-mpv-1.0.1/lib/python3.10/site-packages/mpv.py", line 1606, in register
    types = [MpvEventID.from_str(t) if isinstance(t, str) else t for t in event_types] or MpvEventID.ANY
  File "/nix/store/zr87l99vadmr157sa0vbjvwb5mgkc65c-python3.10-mpv-1.0.1/lib/python3.10/site-packages/mpv.py", line 1606, in <listcomp>
    types = [MpvEventID.from_str(t) if isinstance(t, str) else t for t in event_types] or MpvEventID.ANY
  File "/nix/store/zr87l99vadmr157sa0vbjvwb5mgkc65c-python3.10-mpv-1.0.1/lib/python3.10/site-packages/mpv.py", line 299, in from_str
    return getattr(kls, s.upper().replace('-', '_'))
AttributeError: type object 'MpvEventID' has no attribute 'IDLE'
```

**Comment by jojosch** [comment]
Thats because just like `jellyfin-mpv-shim` the GitHub release does not contain the shaderpack by default, the archive uploaded to [pypi](https://pypi.org/project/plex-mpv-shim/) does. 

The easieast way to fix that in `plex-mpv-shim` would be to use `fetchPypi` instead of `fetchFromGitHub`.

But then the second error `AttributeError: type object 'MpvEventID' has no attribute 'IDLE'` still causes `plex-mpv-shim` to fail on startup. See https://github.com/iwalton3/plex-mpv-shim/issues/72 for the cause and https://github.com/NixOS/nixpkgs/commit/f69011949c17784865bffc3859a3bb5efefacc24 for the commit updating `python-mpv`.

```diff
diff --git a/pkgs/applications/video/plex-mpv-shim/default.nix b/pkgs/applications/video/plex-mpv-shim/default.nix
index 2d0643c4ddb..4d5a9f01fac 100644
--- a/pkgs/applications/video/plex-mpv-shim/default.nix
+++ b/pkgs/applications/video/plex-mpv-shim/default.nix
@@ -1,17 +1,30 @@
-{ lib, buildPythonApplication, fetchFromGitHub, mpv, requests, python-mpv-jsonipc, pystray, tkinter }:
+{ lib, buildPythonApplication, fetchPypi, fetchFromGitHub, mpv, requests, python-mpv-jsonipc, pystray, tkinter }:
 
 buildPythonApplication rec {
   pname = "plex-mpv-shim";
   version = "1.10.3";
 
-  src = fetchFromGitHub {
-    owner = "iwalton3";
-    repo = pname;
-    rev = "v${version}";
-    sha256 = "0hgv9g17dkrh3zbsx27n80yvkgix9j2x0rgg6d3qsf7hp5j3xw4r";
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-GEskhfFxkj59sDt+Ut1UySbU56XNr7i9t0fEaY9/xjc=";
   };
 
-  propagatedBuildInputs = [ mpv requests python-mpv-jsonipc pystray tkinter ];
+  propagatedBuildInputs = [ 
+    # https://github.com/iwalton3/plex-mpv-shim/issues/72
+    (mpv.overrideAttrs (old: rec {
+      version = "0.5.2";
+      src = fetchFromGitHub {
+        owner = "jaseg";
+        repo = "python-mpv";
+        rev = "v${version}";
+        sha256 = "0ffskpynhl1252h6a05087lvpjgn1cn2z3caiv3i666dn1n79fjd";
+      };
+    })) 
+    requests 
+    python-mpv-jsonipc 
+    pystray 
+    tkinter
+  ];
 
   # does not contain tests
   doCheck = false;
```

Maybe there is a nicer way to do this, but this seems to work:

```
$ ./result/bin/plex-mpv-shim
2022-11-06 09:21:45,601 [    INFO] player: Using libmpv1 playback backend.
2022-11-06 09:21:45,623 [    INFO] root: Started GDM service
2022-11-06 09:21:45,624 [    INFO] client: Started HTTP server
2022-11-06 09:21:45,646 [   DEBUG] PIL.PngImagePlugin: STREAM b'IHDR' 16 13
2022-11-06 09:21:45,646 [   DEBUG] PIL.PngImagePlugin: STREAM b'zTXt' 41 1315
2022-11-06 09:21:45,646 [   DEBUG] PIL.PngImagePlugin: STREAM b'bKGD' 1368 6
2022-11-06 09:21:45,646 [   DEBUG] PIL.PngImagePlugin: b'bKGD' 1368 6 (unknown)
2022-11-06 09:21:45,647 [   DEBUG] PIL.PngImagePlugin: STREAM b'pHYs' 1386 9
2022-11-06 09:21:45,647 [   DEBUG] PIL.PngImagePlugin: STREAM b'tIME' 1407 7
2022-11-06 09:21:45,647 [   DEBUG] PIL.PngImagePlugin: b'tIME' 1407 7 (unknown)
2022-11-06 09:21:45,647 [   DEBUG] PIL.PngImagePlugin: STREAM b'IDAT' 1426 612
```

Since I've only got jellyfin, i can't test this any further here - but it's progress :)

**Comment by Myaats** [comment]
> Thats because just like `jellyfin-mpv-shim` the GitHub release does not contain the shaderpack by default, the archive uploaded to [pypi](https://pypi.org/project/plex-mpv-shim/) does.

@jojosch Any feedback on the current changes of this PR, as this seems seems out of scope. ofborg seems to be happy now too.

File: pkgs/development/python-modules/pystray/default.nix [hunk 1]
```diff
@@ -12,8 +22,8 @@ buildPythonPackage rec {
     sha256 = "sha256-8B178MSe4ujlnGBmQhIu+BoAh1doP9V5cL0ermLQTvs=";
   };
 
-  nativeBuildInputs = [ sphinx ];
-  propagatedBuildInputs = [ pillow xlib six ];
+  nativeBuildInputs = [ gobject-introspection sphinx ];
+  propagatedBuildInputs = [ pillow xlib six pygobject3 gtk3 libayatana-appindicator ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pystray/default.nix:26]: Are the scripts in $out/bin now double wrapped?

Comment (Myaats) [reply pkgs/development/python-modules/pystray/default.nix:26]: No, I changed the two applications relying on pystray (which is just a package) to not wrapGapps and added the paths like other python applications does.

**Pull request merged**